### ISSUE TITLE
1.8.4 - Cross-referencing arbitrary locations is not working with large PDF generated documents only

### ISSUE THREAD

**Issue by wtuemura** [sphinx-doc/sphinx#6026]
**Describe the bug**
Cross-referencing arbitrary locations is not working with large PDF generated documents.
https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-referencing-arbitrary-locations

This behavior appears to happen with **PDF documents only**. All the cross-referencing works just fine with HTML pages.

**To Reproduce**
```
.. _mame-commandline-aviwrite:

**-aviwrite** <*filename*>.avi

	Grava todos os dados de áudio e vídeo em formato AVI sem compressão,
	note que a taxa de quadros e a resolução são sempre fixas. Vídeos
	sem compressão ocupam muito espaço assim como, para que a gravação
	ocorra sem problemas é necessário um HDD rápido.
	
	Talvez seja mais prático gravar os seus comandos com
	:ref:`-record <mame-commandline-record>` e
	depois fazer o vídeo com
	:ref:`-aviwrite <mame-commandline-aviwrite>` combinado com
	:ref:`-playback <mame-commandline-playback>` e
	:ref:`-exit_after_playback <mame-commandline-exitafterplayback>`.
```

**Expected behavior**
To generate a link to the proper sections.

**Your project**
Download the last source update from the main page.
https://github.com/wtuemura/mamedoc

**Screenshots**
Sphinx **1.8.3**, we have the cross reference working.
![sphinx-183](https://user-images.githubusercontent.com/44546769/52293672-95e52d80-295e-11e9-81f6-9c777dd521de.png)

Sphinx **1.8.4**, with no changes in my source, the cross reference doesn't work anymore.
![sphinx-184](https://user-images.githubusercontent.com/44546769/52293765-cdec7080-295e-11e9-87cf-82708027eaae.png)

There are parts of the document that got cross-referenced and others that don't for unknown reason.

Also noticed over 1000+ warnings like ``LaTeX Warning: Hyper reference `initialsetup/compilingmame:compiling-mame' on page 7 undefined on input line 346.`` Sphinx complains about it but looking in to source there are nothing wrong with.

**initialsetup/compilingmame.rst**:
```
.. _compiling-mame:

Compilando o MAME
=================
```
**initialsetup/installingmame.rst**:
```
Caso você escolha a opção de compilar você mesmo o MAME, consulte a
seção :ref:`compiling-mame` para mais detalhes.
```
**initialsetup/crosscompilemame.rst**:
```
Assim como na compilação nativa, você pode adicionar a opção **-j** no
final do comando visando acelerar o processo de compilação usando os
núcleos do seu processador como já foi explicado com mais detalhes no
capítulo :ref:`compiling-mame`: ::
```

**Environment info**
- OS: **Debian 9.7 x64**
- Python version: **2.7.13**
- Sphinx version: **1.8.4**
- Sphinx extensions:  **sphinx.ext.autodoc, sphinx.ext.mathjax, sphinx.ext.viewcode, edit_on_github, sphinxcontrib.rsvgconverter**.
- Extra tools: **sphinx_rtd_theme** for HTML generation.

**Additional context**
Installed with `pip install -U Sphinx`

Thank you.

**Comment by jfbu** [comment]
Caused by #6006. Thanks for report.

**Comment by jfbu** [comment]
small project showing the issue: [6026project.zip](https://github.com/sphinx-doc/sphinx/files/2834080/6026project.zip)

**Comment by tk0miya** [comment]
Fixed by #6051.
Thank you for reporting.

**Comment by wtuemura** [comment]
Confirmed @tk0miya 
完璧ですよねぇ。

Thank you.
